src/preprocessing/passes/symmetry_detect.cpp

```diff
@@ -0,0 +1,456 @@
+  }
+  // If node is a constant
+  else if (node.isConst())
+  {
+    p.d_term = node;
+    d_term_partition[node] = p;
+    return p;
+  }
+
+  // Children of node
+  vector<Node> children;
+  // Partitions of children
+  vector<Partition> partitions;
+
+  // Get all children of node
+  Trace("sym-dt") << "[sym-dt] collectChildren for: " << node
+                  << " with operator " << node.getKind() << endl;
+  collectChildren(node, children);
+  Trace("sym-dt") << "[sym-dt] children: " << children
+                  << endl;
+
+  // Create partitions for children
+  for (vector<Node>::iterator children_it = children.begin();
+       children_it != children.end();
+       ++children_it)
+  {
+    partitions.push_back(findPartitions(*children_it));
+  }
+
+  Trace("sym-dt") << "----------------------------- Start processing "
+                     "partitions -------------------------------"
+                  << endl;
+
+  PartitionTrie pt;
+  unordered_set<Node, NodeHashFunction> vars;
+
+  if (theory::quantifiers::TermUtil::isComm(node.getKind()))
+  {
+    // Start processing the singleton partitions and collect variables
+    processSingletonPartitions(partitions, vars);
+  }
+  else
+  {
+    // Get all the variables from the partitions
+    getVariables(partitions, vars);
+  }
+
+  // Build the partition trie
+  for (unordered_set<Node, NodeHashFunction>::iterator vars_it = vars.begin();
+       vars_it != vars.end();
+       ++vars_it)
+  {
+    pt.addNode(*vars_it, partitions);
+  }
+
+  // Get the new partition
+  pt.getNewPartition(p, pt);
+
+  // Reconstruct the node
+  Trace("sym-dt") << "[sym-dt] Reconstructing node: " << node << endl;
+  p.d_term = node;
+  d_term_partition[node] = p;
+  printPartition(p);
+  return p;
+}
+
+void SymmetryDetect::matches(vector<Partition>& partitions,
+                             map<Node, Node>& subvar_to_var,
+                             map<Node, Node>& subvar_to_expr)
+{
+  Trace("sym-dt")
+      << "[sym-dt] Start testing if singleton partitions can be merged!"
+      << endl;
+
+  if (!subvar_to_expr.empty())
+  {
+    // Replacement variables
+    vector<Node> repls;
+    // Variables that have been merged
+    unordered_set<Node, NodeHashFunction> merged;
+    // Put the variable in repls
+    repls.push_back((subvar_to_expr.begin())->first);
+
+    for (map<Node, Node>::iterator expr_it1 = subvar_to_expr.begin();
+         expr_it1 != subvar_to_expr.end();
+         ++expr_it1)
+    {
+      // Skip expr_it1->first, if it has been merged
+      if (merged.find(expr_it1->first) != merged.end())
+      {
+        continue;
+      }
+
+      Partition p;
+      vector<Node> subs;
+      vector<Node> part;
+      Node fst_var = subvar_to_var.find(expr_it1->first)->second;
+
+      part.push_back(fst_var);
+      subs.push_back(fst_var);
+      merged.insert(expr_it1->first);
+      p.d_var_to_subvar[fst_var] = expr_it1->first;
+      Node sub_expr1 =
+          (expr_it1->second)
```

Comment:
It looks like there is only a single substitution, so you should be able to do: `(expr_it1->second).substitute(fst_var, (subvar_to_expr.begin())->first);` so you don't need the `subs` vector and the `repl` vector. Note: I seem to recall that you might have to manually convert the arguments to TNodes: `(expr_it1->second).substitute(TNode(fst_var), TNode((subvar_to_expr.begin())->first));`